### PR TITLE
Create tests for views and models

### LINKED ISSUE #23: Create tests
- [x] Read about tests [here](http://www.drdobbs.com/testing/unit-testing-with-python/240165163)
- [x] Also there's a good link about test's in django docs [here](https://docs.djangoproject.com/en/1.10/topics/testing/)
- [x] Create tests for models
- [x] Create tests for views
- [x] Integrate your app with Travis CI and run test at Travis' script

As an example you can use [this](https://github.com/kpi-web-guild/django-girls-blog-DrEdi/blob/master/vagga.yaml), but try to do it by yourself :new_moon_with_face:

### PR DESCRIPTION
**Pull request by makkaronis4e**
To be honest i've copied tests from previous student, but adopted it to my project.
Also i've read all articles. As i understood some tests didnt pass because there was no posts: 

```
FAIL: test_index_view_rendering (blog.tests.test_views.ViewsTest)
Testing main page for render needed posts.
```
  Resolve #23

### CODE REVIEW

**Review by webknjaz — CHANGES_REQUESTED** [review]
Add running tests to Travis.

File: blog/tests/test_models.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    def setUp(self):
+        """Prepare data for testing."""
+        self.user = User.objects.create(username='testuser')
+        self.test_post = Post.objects.create(author=self.user, title='Test', text='superText')
+
+    def test_post_rendering(self):
+        """Post is rendered as its title."""
+        self.assertEqual(str(self.test_post), self.test_post.title)
+
+    @patch('django.utils.timezone.now', lambda: datetime(day=1, month=4, year=2016,
```
Comment (webknjaz) [inline blog/tests/test_models.py:24]: Do you know about mock, patch and why they are needed?

File: blog/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+class ViewsTest(TestCase):
+    """Testing class for views."""
+
+    USERNAME = 'mak'
+    PASSWORD = 'MAKkaronis4e'
+
+    def setUp(self):
+        """Prepare data for testing."""
+        self.client = Client()
+        self.user = User.objects.create(username='mak', password='MAKkaronis4e', is_superuser=True, is_staff=True,
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Why did you hardcode creds?

File: .travis.yml [hunk 3]
```diff
@@ -3,3 +3,4 @@ python: 3.6
 cache: pip
 before_install: pip install -U pre-commit
 script: pre-commit run --all-files
+script: python manage.py test
```
Comment (webknjaz) [inline .travis.yml]: It's a duplicate key in dict, which is invalid. See docs for Travis

**Review by webknjaz — CHANGES_REQUESTED** [review]

**Review by webknjaz — CHANGES_REQUESTED** [review]
What does "some" mean?

File: blog/tests/test_views.py [hunk 4]
```diff
@@ -45,10 +55,10 @@ def test_index_view_rendering(self):
             self.assertNotContains(response, future_post)
         with patch('django.utils.timezone.now', lambda: datetime(day=1, month=4, year=3016, tzinfo=tz)):
             response = self.client.get(reverse('post_list'))
-            self.assertListEqual(list(response.context['posts']), [post, past_post, future_post])
+            self.assertListEqual(list(response.context['posts']), [future_post, post, past_post])
         response = self.client.get(reverse('post_list'))
         self.assertEqual(200, response.status_code)
-        self.assertTemplateUsed(response, 'main/index.html')
+        self.assertTemplateUsed(response, 'blog/base.html')
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Why is it base? I see nothing regarding base in views. This test is invalid.

File: blog/tests/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+"""This package contains tests for project."""
```
Comment (webknjaz) [inline blog/tests/__init__.py]: Don't ever refer to self via "This ...". It is obvious that if the docstring refers to the scope where it resides.
Writing explanations this way is useless.

File: blog/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        response = self.client.get(reverse('post_new'))
+        self.assertEqual(302, response.status_code)
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
+        self.assertTrue(authorization)
+        response = self.client.post(reverse('post_new'), {'author': self.user, 'title': 'Test', 'text': 'superText', },
+                                    follow=True)
+        self.assertEqual(200, response.status_code)
+        self.assertRedirects(response, reverse('post_detail', kwargs={'pk': 1}))
+
+    def test_post_edit(self):  # and this one
```
Comment (webknjaz) [inline blog/tests/test_views.py]: You test the view, but also check post edit flow.

Also check that:
1) the post actually appeared on the page
2) if you **EDIT** existing post, changes are reflected

File: blog/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        self.assertContains(response, past_post)
+        self.assertNotContains(response, future_post)
+
+    def test_detail_view(self):
+        """Testing detail page when post is not exist and when it exists."""
+        response = self.client.get(reverse('post_detail', kwargs={'pk': 1}))
+        self.assertEqual(404, response.status_code)
+        post = Post.objects.create(author=self.user, title='Test', text='superText')
+        response = self.client.get(reverse('post_detail', kwargs={'pk': post.pk}))
+        self.assertEqual(200, response.status_code)
```
Comment (webknjaz) [inline blog/tests/test_views.py:71]: Did the post actually appear on the web page?

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: blog/tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            self.assertNotContains(response, future_post)
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=4, year=3016, tzinfo=tz)):
+            response = self.client.get(reverse('post_list'))
+            self.assertListEqual(list(response.context['posts']), [future_post, post, past_post])
+        response = self.client.get(reverse('post_list'))
+        self.assertEqual(200, response.status_code)
+        self.assertTemplateUsed(response, 'blog/post_list.html')
+        self.assertContains(response, post)
+        self.assertContains(response, past_post)
+        self.assertNotContains(response, future_post)
```
Comment (webknjaz) [inline blog/tests/test_views.py]: This check is non-deterministic, it will fail in about 998 years.

Comment (webknjaz) [reply blog/tests/test_views.py]: It could be moved into second `with` block.

File: blog/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            self.assertNotContains(response, post)
+            self.assertNotContains(response, future_post)
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=4, year=2016, tzinfo=tz)):
+            response = self.client.get(reverse('post_list'))
+            self.assertListEqual(list(response.context['posts']), [post, past_post])
+            self.assertNotContains(response, future_post)
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=4, year=3016, tzinfo=tz)):
+            response = self.client.get(reverse('post_list'))
+            self.assertListEqual(list(response.context['posts']), [future_post, post, past_post])
+        response = self.client.get(reverse('post_list'))
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Refactor starting this line.

File: blog/tests/test_views.py (hunk 7 above)
Comment (webknjaz) [reply blog/tests/test_views.py:71]: So?

File: blog/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    def test_post_new_view(self):  # so only the tests with login are broken
+        """Testing new post view before and after login."""
+        response = self.client.get(reverse('post_new'))
+        self.assertEqual(302, response.status_code)
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
+        self.assertTrue(authorization)
+        response = self.client.post(reverse('post_new'), {'author': self.user, 'title': 'Test', 'text': 'superText', },
+                                    follow=True)
+        self.assertEqual(200, response.status_code)
+        self.assertRedirects(response, reverse('post_detail', kwargs={'pk': 1}))
```
Comment (webknjaz) [inline blog/tests/test_views.py]: You could also check that the post actually appears on the web page

File: blog/tests/test_views.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        self.assertTrue(authorization)
+        response = self.client.get(reverse('post_edit', kwargs={'pk': 1}))
+        # This time we'are logged but post is not exist
+        self.assertEqual(404, response.status_code)
+        post = Post.objects.create(author=self.user, title='Test', text='superText')
+        response = self.client.get(reverse('post_edit', kwargs={'pk': post.pk}))
+        self.assertEqual(200, response.status_code)
+        response = self.client.post(reverse('post_edit', kwargs={'pk': 1}), {'author': self.user, 'title': '',
+                                    'text': 'asdasdasd', }, follow=True)
+        self.assertEqual(200, response.status_code)
```
Comment (webknjaz) [inline blog/tests/test_views.py]: Also check the web page contents

**Review by webknjaz — CHANGES_REQUESTED** [review]

**Review by webknjaz — CHANGES_REQUESTED** [review]
You should react to all comments

File: blog/tests/test_views.py (hunk 6 above)
Comment (makkaronis4e) [reply blog/tests/test_views.py]: Done

File: blog/tests/test_views.py (hunk 7 above)
Comment (makkaronis4e) [reply blog/tests/test_views.py:71]: Done

File: blog/tests/test_views.py (hunk 10 above)
Comment (makkaronis4e) [reply blog/tests/test_views.py]: In new post we dont create a post, we just look if 'new post' form works correct

File: blog/tests/test_models.py (hunk 1 above)
Comment (makkaronis4e) [reply blog/tests/test_models.py:24]: Yep

File: blog/tests/test_views.py (hunk 10 above)
Comment (webknjaz) [reply blog/tests/test_views.py]: No, you create post a few lines above.

File: blog/tests/test_views.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        self.assertTrue(authorization)
+        response = self.client.get(reverse('post_edit', kwargs={'pk': 1}))
+        # This time we'are logged but post is not exist
+        self.assertEqual(404, response.status_code)
+        post = Post.objects.create(author=self.user, title='Test', text='superText')
+        response = self.client.get(reverse('post_edit', kwargs={'pk': post.pk}))
+        self.assertEqual(200, response.status_code)
+        response = self.client.post(reverse('post_edit', kwargs={'pk': 1}), {'author': self.user, 'title': 'Test',
+                                    'text': 'asdasdasd', }, follow=True)
+        self.assertContains(response, post)
```
Comment (webknjaz) [inline blog/tests/test_views.py]: No, check that it contains changed text there. Because now you only check that the model is still the same

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: blog/tests/test_views.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        self.assertEqual(302, response.status_code)
+        authorization = self.client.login(username=self.USERNAME, password=self.PASSWORD)
+        self.assertTrue(authorization)
+        response = self.client.get(reverse('post_edit', kwargs={'pk': 1}))
+        # This time we'are logged but post is not exist
+        self.assertEqual(404, response.status_code)
+        post = Post.objects.create(author=self.user, title='Test', text='superText')
+        response = self.client.get(reverse('post_edit', kwargs={'pk': post.pk}))
+        self.assertEqual(200, response.status_code)
+        response = self.client.post(reverse('post_edit', kwargs={'pk': 1}), {'author': self.user, 'title': 'Test',
```
Comment (webknjaz) [inline blog/tests/test_views.py:98]: in fact, you don't have to change author and title when testing change of contents only.

File: blog/tests/test_views.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=1, year=2016, tzinfo=tz)):
+            response = self.client.get(reverse('post_list'))
+            self.assertListEqual(list(response.context['posts']), [past_post])
+            self.assertNotContains(response, post)
+            self.assertNotContains(response, future_post)
+        with patch('django.utils.timezone.now', lambda: datetime(day=1, month=4, year=2016, tzinfo=tz)):
+            response = self.client.get(reverse('post_list'))
+            self.assertListEqual(list(response.context['posts']), [post, past_post])
+            self.assertTemplateUsed(response, 'blog/post_list.html')
+            self.assertContains(response, post)
```
Comment (webknjaz) [inline blog/tests/test_views.py:56]: You check this two lines above

File: blog/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+            self.assertContains(response, past_post)
```
Comment (webknjaz) [inline blog/tests/test_views.py:57]: This as well

File: blog/tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+            self.assertNotContains(response, future_post)
```
Comment (webknjaz) [inline blog/tests/test_views.py:58]: and this

**Review by webknjaz — APPROVED** [review]

**Pull request merged**